LE Audio: Upstream BAPS

Thalley opened the pull request:
Upstreams BAPS which includes

- BAP Unicast Server
    - ASCS
    - PACS
- BAP Unicast Client
- BAP Broadcast Source
- BAP Broadcast Sink  

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41190

Thalley committed:
Bluetooth: Audio: Add BAP unicast server support

Add the Basic Audio Profile (BAP) unicast server
functionality. This allows a device to act as the
unicast server role, which can accept unicast streams
initiated by a unicast client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Tests: Fix ctrl iso audio dependency

Change the dependency from BT_AUDIO to ISO.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BAP unicast client support

Add the BAP unicast client implementation. This role
can discover BAP unicast server services and initiate
BAP audio streams.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BAP unicast audio BSIM tests

Add initial babblesim tests for unicast audio.
Due to lack of ISO support in the babblesim, these
won't attempt to actual initate audio streams.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Samples: Add unicast audio client and server samples

Adds Basic Audio Profile (BAP) unicast server and client samples.

These are the barebones versions of what is needed to scan/advertise
for audio and setup a stream using the mandatory LC3 preset defined by
the BAP spec.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BAP broadcast source support

Add the BAP broadcast source implementation. This role
allows a device to broadcast ISO data.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add BAP broadcast sink support

Add support for the BAP broadcast sink role. This role
allows a device to sync to a broadcast ISO stream.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add broadcast audio BSIM tests

Add broadcast babblesim test cases that tests basically
functionality. Due to the lack of ISO support in the
controller, this won't actually start any audio streams.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: shell: Add LE Audio shell commands

Adds audio.c which covers all the LE Audio shell commands.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo commented on doc/reference/bluetooth/shell/audio.rst:
```diff
@@ -0,0 +1,479 @@
+Init
+****
+
+This command register local PAC records which are necessary to be able to
```
Even though the commands are listed above in the file, I think it makes sense to mention them (again) here, instead of saying "this command" without actually specifying which command "this" is.

Proposal here, for init:
The "init" command" registers local PAC records [...]  (note also the "s" in registers)

Proposal for discover (below):
Once connected, the "discover" command is used to discover [...]

Ditto for remaining commands

asbjornsabo commented on doc/reference/bluetooth/shell/audio.rst:
```diff
@@ -0,0 +1,479 @@
+Select preset
+*************
+
+This command can be used to either print the default preset configuration or set
```
The "preset" command [...]

Thalley commented on doc/reference/bluetooth/shell/audio.rst:
```diff
@@ -0,0 +1,479 @@
+   data #0: type 0x01 len 1
+   data #1: type 0x02 len 1
+   data #2: type 0x03 len 4
+   00000000: 00 00 00                                         |...              |
```
That's just how it is shown when you print data with `shell_hexdump` :)

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1699 @@
+	BT_CODEC_QOS_INOUT
+};
+
+/** @brief Codec QoS direction */
```
Framing, not direction

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1699 @@
+	BT_CODEC_QOS(BT_CODEC_QOS_OUT, _interval, BT_CODEC_QOS_UNFRAMED, \
+		     BT_CODEC_QOS_2M, _sdu, _rtn, _latency, _pd)
+
+/** @def BT_CODEC_QOS_OUT_UNFRAMED
```
INOUT

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1699 @@
+	BT_CODEC_QOS(BT_CODEC_QOS_OUT, _interval, BT_CODEC_QOS_FRAMED, \
+		     BT_CODEC_QOS_2M, _sdu, _rtn, _latency, _pd)
+
+/** @def BT_CODEC_QOS_OUT_FRAMED
```
INOUT

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1699 @@
+	BT_AUDIO_SOURCE = 0x02,
+} __packed;
+
+#define BT_AUDIO_CONTENT_UNSPECIFIED     BIT(0)
```
I guess these should _CONTEXT_, not _CONTENT_

Also, these overlap with the codec defines in line 123 - 189 - do we need both?
(Or are they actually separate entities in the spec?)

Thalley replied:
I can't find them in any specification, so I decided to remove them, as we have the context values instead.

asbjornsabo replied:
There has been a lot of confusion between "content" and "context" - "context" is the correct one.
(I believe there are no content values.)

asbjornsabo commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1699 @@
+/** @brief Create audio unicast group.
+ *
+ *  Create a new audio unicast group with one or more audio streams as a
+ *  central. Streams in a unicast group shall share the same interval, framing
```
Previously, "client" and "server" has been used.  Should this "central" be "server"?
(Ditto in next function.)

Thalley replied:
Changed to `client`. Should it be more specific and be `unicast client`?

asbjornsabo replied:
"client" is understandable - the unicast client is the only client we have (here), so it must be that one.
"unicast client" is even better though, as that is the term elsewhere, so I would prefer that.

asbjornsabo commented on include/bluetooth/audio/capabilities.h:
```diff
@@ -0,0 +1,260 @@
+
+	/** @brief Capability Start callback
+	 *
+	 *  Start callback is called whenever an Audio Stream is about to be
```
"to start streaming" or "to be started streaming"

Thalley replied:
Fixed this and a few similar oddities in the descriptions.

asbjornsabo commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -0,0 +1,638 @@
+	}
+}
+
+static bool check_audio_support(struct bt_data *data, void *user_data)
```
The name of the function does not seem very descriptive for a function that actually connects ...
"connect_if_audio_support"?

asbjornsabo commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -0,0 +1,638 @@
+	if (err != 0) {
+		return;
+	}
+	printk("Channel configured\n");
```
"Channel" vs "stream"

asbjornsabo commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -0,0 +1,638 @@
+	if (err != 0) {
+		return;
+	}
+	printk("Channel QoS Set\n");
```
Ditto

asbjornsabo commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1953 @@
+	ep = &ase->ep;
+
+	/* If the ASE_ID  written by the client represents a Sink ASE, the
+	 * server shall not accept the Receiver Start Ready operation for that
```
This is the "stop" operation, not the "receiver start ready".
Seems like a copy and paste error.

Thalley replied:
Seems like it, indeed. Updated to `stop`.

Thalley commented on samples/bluetooth/unicast_audio_server/boards/nrf52840dk_nrf52840.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_BT_CTLR_ADV_EXT=y
```
So it doesn't build for the nrf52840?

asbjornsabo replied:
(This is for the shell without this PR.)
No, if I remember correctly, I had to enable the configs in the project file.  Did not have time to look more into it.
Anyway - off-topic here.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1953 @@
+		}
+
+		if (err != 0) {
+			uint8_t reason = BT_ASCS_REASON_CODEC_DATA;
```
That's implementation specific. The spec doesn't specify any reason for that purpose, and the others aren't any better, so I think this is the best response in this case.

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1953 @@
+							&ase->ep.codec, &stream,
+							&ase->ep.qos_pref);
+		} else {
+			err = -EOPNOTSUPP;
```
This is just a local value, and yes that is the best to use here as well, although it needs to be `ENOTSUP` instead (updated).

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1953 @@
+		struct bt_ascs_ase *ase;
+		int err;
+
+		if (buf->len < sizeof(*cfg)) {
```
Yes, because we do `net_buf_simple_pull_mem` in a loop

asbjornsabo replied:
First, there is a check that the buf contains at least req->num_ases blocks of cfg-size.
Then there is a loop, req->num_ases times, which each pull one cfg-size block.

I still do not see how that second operation can fail.  But then again, the check does no harm, I just can not see it is necessary.

Thalley replied:
It isn't very clear from the code, but in line 1090 there's a call to `ase_config` which then extracts the variable length input for each item in the array.

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -0,0 +1,1100 @@
+	const struct bt_codec_qos_pref *qos_pref = &stream->ep->qos_pref;
+
+	if (qos_pref->latency < qos->latency) {
+		BT_DBG("Latency %u higher than preferred max %u",
```
I agree that this could be done differently. It's not a choice I've made (existing code), but I do see the point of not letting the upper layers set values that are outside the preference. We could let this be up to the upper layers completely though.

asbjornsabo replied:
We'll leave it for now.

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -0,0 +1,1956 @@
+static void unicast_client_ep_streaming_state(struct bt_audio_ep *ep,
+					      struct net_buf_simple *buf)
+{
+	struct bt_ascs_ase_status_enable *enable;
```
Agreed. Added new structs for the streaming and disabling states. 

The variable's purpose is just to extract memory yes, as well as providing a shorthand way of verifying length. It could be removed and replaced by `sizeof(struct bt_ascs_ase_status_enable)`.

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -0,0 +1,1956 @@
+	       bt_audio_ep_state_str(ep->status.state),
+	       bt_audio_ep_state_str(state));
+
+	/* TODO: Verify state changes with keeping in mind that broadcast
```
Old TODO :)

asbjornsabo reviewed (approved):
I have now spent the best part of three days reviewing the core parts of this PR, without finding any significant issues.
I have also reviewed this code in a number of PRs as it has been developed, changed and maintained in the topic branch over a long period of time.
In total, I believe that this sound and good code that should be merged to main so that we can continue the development there.

Thalley commented:
@jhedberg @alwa-nordic Any chance you guys can take a look at this, so we have a chance to get it merged before Friday?

jhedberg commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1690 @@
+	 *  This shall be set to BT_CODEC_QOS_OUT for broadcast sources, and
+	 *  shall be set to BT_CODEC_QOS_IN for broadcast sinks.
+	 */
+	uint8_t  dir;
```
Doesn't necessarily have to be done in this PR, but unless the ordering of these struct members is critical, you could reduce the size of this struct by 32 bits by moving the `dir` member to the same 32-bit block that's occupied by `rtn` and `latency` (then there wouldn't be any unnecessary padding).

jhedberg commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1690 @@
+enum bt_audio_pac_type {
+	BT_AUDIO_SINK = 0x01,
+	BT_AUDIO_SOURCE = 0x02,
+} __packed;
```
The `__packed` seems unnecessary, since you're nowhere referring directly to the enum type, rather only assigning the enum values to `uint8_t` typed variables.

jhedberg commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1690 @@
+
+struct bt_audio_discover_params {
+	/** Capabilities type */
+	uint8_t  type;
```
Here's one example of ignoring the enum type itself and just going with an `uint8_t`. If you really want to use a packed enum, take a look at `bt_conn_state_t`. Otherwise, just remove the `__packed` from the definition since it doesn't give you any benefit.

jhedberg commented on include/bluetooth/audio/capabilities.h:
```diff
@@ -0,0 +1,260 @@
+#endif
+
+/* Get list of capabilities by type */
+sys_slist_t *bt_audio_capability_get(uint8_t type);
+
+#define BT_AUDIO_CAPABILITY_UNFRAMED_SUPPORTED     0x00
+#define BT_AUDIO_CAPABILITY_UNFRAMED_NOT_SUPPORTED 0x01
```
This would seem to be a different kind of capability type, which would speak in favour of using the full enum type instead of `uint8_t`, to make it clear exactly what value domain you're dealing with.

Thalley replied:
I moved the values into an enum, but will not use the enum as the values are assigned to a __packed struct.

jhedberg commented on subsys/bluetooth/audio/broadcast_sink.c:
```diff
@@ -0,0 +1,1033 @@
+		}
+	}
+
+	if (sizeof(len) > buf->len) {
```
The convention in most of the tree (and which I also find more intuitive to read) is to have the variable on the left side and the constant on the right side of compatisons.

jhedberg commented on subsys/bluetooth/audio/broadcast_sink.c:
```diff
@@ -0,0 +1,1033 @@
+	 * every LTV entry, which would simply be too much for handling
+	 * broadcasted BASEs
+	 */
+	while (ltv_buf.len != 0) {
```
..like you do here. So the code is also inconsistent with itself.

rymanluk commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -0,0 +1,1952 @@
+
+	BT_DBG("ep %p len %u codec %p", ep, len, codec);
+
+	if (!len) {
```
same as above - if metadata change, what will happen with previously set metadata?

rymanluk commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -0,0 +1,1952 @@
+	/* Read ASE instances */
+	if (unicast_client_ase_discover(conn, params) < 0) {
+		BT_ERR("Unable to read ASE");
+		goto fail;
```
looks like fail label in this function is not needed.

rymanluk reviewed (approved):
Looks good. Couple of minor comments, but this PR is big and tested on IOPs so it is better to merge it and do fixes on top of it.

Thalley commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1689 @@
+	union {
+		void *group;
+		struct bt_audio_unicast_group *unicast_group;
+		struct bt_audio_broadcast_source *broadcast_source;
```
Not really. 

One shame about moving the code in new commits, is that all the reasoning from the earlier commit messages are lost :D 

This is fairly old code, where the broadcast sink and sources were significantly different. The code works, but it's far from optimal. This struct can be further optimized for each role.

rymanluk replied:
Ok

Thalley commented on include/bluetooth/audio/audio.h:
```diff
@@ -1,9 +1,1689 @@
+	sys_snode_t node;
+
+	union {
+		void *group;
```
It is just a way to generically check the value of any of the possible groups. It is used e.g. in `bt_audio_broadcast_sink_setup_stream`: 
```
	if (stream->group != NULL) {
		BT_DBG("Stream %p already in group %p", stream, stream->group);
		return -EALREADY;
	}
```

rymanluk replied:
Ok I see. Thanks.

Thalley commented on subsys/bluetooth/audio/unicast_client.c:
```diff
@@ -0,0 +1,1952 @@
+	codec->vid = vid;
+	codec->data_count = 0;
+
+	if (!len) {
```
Added a memset to clear it.

Thalley committed:
Bluetooth: Audio: Add BT_WARN for missing stream callbacks

If any specific stream callback is attempted
to be called but is missing, a BT_WARN log statement
is added.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1961 @@
+
+	req = net_buf_simple_pull_mem(&buf, sizeof(*req));
+
+	BT_DBG("conn %p attr %p buf %p len %u offset %u op %s (0x%02x)", conn,
```
I guess you might remove offset from log as it will always be 0.

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -0,0 +1,1961 @@
+		ret = ascs_release(ascs, &buf);
+		break;
+	default:
+		ascs_cp_rsp_add(0x00, req->op, BT_ASCS_RSP_NOT_SUPPORTED, 0);
+		BT_DBG("Unknown opcode");
+		len = BT_GATT_ERR(BT_ATT_ERR_NOT_SUPPORTED);
```
Does it mean ASE CP notification will be sent as well along with ATT error (as write response)?

MariuszSkamra replied:
Ah, I guess it's because ATT Write Without Response might be used. Isn't it?

Thalley replied:
> Ah, I guess it's because ATT Write Without Response might be used. Isn't it?

Correct :)

MariuszSkamra reviewed (approved):
LGTM

Thalley commented:
@dkalowsk Can this be merged post feature freeze? It doesn't change any existing code, and is off by default, and is a blocking feature a lot of future work for Bluetooth LE Audio.

dkalowsk commented:
@Thalley having looked it through, I don't claim to know all of it.   Just curious on my side, any specific reason it missed the 3.0 tag cutoff?  As to your question on merging, I don't see any reason why it cannot, although I believe we need to check with the TSC on this, is that right @nashif ?

Thalley commented:
> @Thalley having looked it through, I don't claim to know all of it. Just curious on my side, any specific reason it missed the 3.0 tag cutoff? As to your question on merging, I don't see any reason why it cannot, although I believe we need to check with the TSC on this, is that right @nashif ?

It didn't receive the necessary approvals by Friday, but got it today.

Thalley commented:
@dkalowsk @jhedberg @carlescufi since this was approved by the TSC yesterday, can we get it merged today please?
